### DISCUSSION TITLE
React-select wrapped in Controller breaks onChange

### DISCUSSION THREAD

**Discussion by alvamanu** [react-hook-form/react-hook-form#2433] [Q&A] [answered]
I'm trying to implement a Select All feature for my react-select field. I've already built this feature and it works without react-hook-form, but I can't accomplish this if the component is wrapped in Controller. I'm guessing the problem is Controller is overriding handleChange for the component inside of it. 

I've built a codesandbox with both scenarios: with controller and without controller.

https://codesandbox.io/s/gracious-flower-zu4mw?file=/src/App.js

I believe the issue could be solved if I would programmatically invoke an onChange on the Controller. I even tried to add a ref to the component, but it doesn't allow it. Any help is greatly appreciated.

**Comment by bluebill1049** [comment]
<img width="1282" alt="Screen Shot 2020-07-31 at 11 30 41 am" src="https://user-images.githubusercontent.com/10513364/88990422-45674a00-d321-11ea-8941-4c64b4e19414.png">

it's broken with your CSB.

**Comment by bluebill1049** [comment]
let's continue the chat in here please @manualva7 this is most likely an expected behavior.

**Comment by bluebill1049** [comment]
@manualva7 btw the usage looks incorrect, probably is the reason why it's not working. take a look at my codesandbox with `render` prop, see if this will fix the issue.

**Comment by bluebill1049** [comment]
here is my codesandbox: https://codesandbox.io/s/react-hook-form-v6-controller-qsd8r

**Reply by bluebill1049** [reply]
in the `render` prop: you don't need to set up `handleChange` and `setValue`, controller should do most of the work for you.

**Reply by bluebill1049** [reply]
<img width="607" alt="Screen Shot 2020-07-31 at 1 25 11 pm" src="https://user-images.githubusercontent.com/10513364/88996698-4607dc80-d331-11ea-9ead-561d1d4c81d8.png">

**Reply by bluebill1049** [reply]
render prop

<img width="616" alt="Screen Shot 2020-07-31 at 1 26 47 pm" src="https://user-images.githubusercontent.com/10513364/88996773-7ea7b600-d331-11ea-9099-36778910d133.png">

**Comment by alvamanu** [comment, marked as answer]
render doesn't work for me. It's giving me an error. I don't know if it's the version I'm using.

I passed the content of "as" as a function.
https://codesandbox.io/s/objective-robinson-g8qrc

It works in sandbox however, it's too buggy for the application I'm creating.
I need to do a refactor for validation on my entire application. I would like to keep using this plugin, but Controller is not going to work for me. Can you point out how to validate state instead of input? Thanks

**Reply by bluebill1049** [reply]
if `Controller` is not going to work for you, perhaps you will have to do the value update manually. register the input at `useEffect` and update data with `setValue`?

**Reply by alvamanu** [reply]
Using `setValue` hook from `useForm` works. Thanks!

**Comment by alvamanu** [comment]
The problem is the react-select being wrapped in Controller. Right now I'm using Autocomplete with static options, with async options and now I'm using it with chips (multiple values). And on those multiple values I'm adding an add all options button. The amount of types of fields is complex for the app I'm working on.

I need to validate state instead of input. Is there a way to validate bypassing "ref={register}" for each input? Or is "ref={register}" essential for all methods of validation in react-hook-form?

**Reply by bluebill1049** [reply]
do you have a simple coddsandbox to reproduce the validation you want to have?  I think it's matter of how you want to validate your inputs and whether you use build-in or schema validation.

**Reply by alvamanu** [reply]
The problem with validating value directly inside an input, taking into account UI libraries will be used, is that: 

1) the input is not necessarily displaying what's going to be submitted.
2) the field is not necessarily using input/value method.

I understand the reason for Controller, but there will be times when the developer will want to invoke onChange programmatically. Given Controller doesn't accept ref prop I believe this is not possible.

One of the sure things about react form fields is that their state will always be present. No matter how you decide to manipulate it, state will always be key/value, unlike an input's value. A "single source of truth" if you can call it that, like redux's store.

And this should always be taken into consideration because when a phone field has a mask, you can't validate that input with a maxLength of 10 because masks add more characters to the field's value. So I think the best way to go is to validate state.

I had to hack my way through this. I've built a workaround for this issue. I've bound the name of the react-select component to a "twin" hidden input. That hidden input's value (with the same name of its react-select component) is the state of the component that is bound to it. ref={register} is then passed to that hidden input. 

I'm currently building a simple form that would only validate state. I'll let you know when I'm done.

**Reply by bluebill1049** [reply]
Thanks for the detailed report above, i will see what you going to have in the CSB and potentially provide a simpler solution.

> I understand the reason for Controller, but there will be times when the developer will want to invoke onChange programmatically. Given Controller doesn't accept ref prop I believe this is not possible.

Why not use `setValue`? to programmatically update the input?

**Reply by alvamanu** [reply]
I'm using setValue.

So I've built the "Add all options" button. It grabs the whole array of objects and plops it into the fields state, like so: 

setValue(options)

Super simple. But since onChange in Controller isn't invoked, I can't see the values inside the field. They're in state, just not in the DOM.

This is what I'm dealing with

![Screen Shot 2020-08-03 at 9 03 13 PM](https://user-images.githubusercontent.com/41272874/89251794-cac25700-d5cc-11ea-8d11-a5de0ada5e9a.png)

**Reply by bluebill1049** [reply]
can you share me this codesandbox? I would love to have a play.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/react-hook-form-v6-controller-6zwie?file=/src/index.js
works for me.

**Reply by wasiqkhan** [reply]
@alvamanu could you share the codesandbox? I am building something similar to what you are building.

**Comment by bluebill1049** [comment]
hey @manualva7 

sorry for the late notice... thanks very much for the support this project. 🙏  

<img width="427" alt="Screen Shot 2020-08-04 at 5 10 39 pm" src="https://user-images.githubusercontent.com/10513364/89263936-745d2300-d675-11ea-8ccf-f18c1431924a.png">

cheers
bill

**Reply by alvamanu** [reply]
no problem. Looking forward for more awesome projects

**Comment by alvamanu** [comment]
Ok I managed to use setValue and it works great. The name got me confused. I thought you were saying `setValue` as in useState. Now I'm seeing is a hook from `useForm`. 

Thanks for your help! This is solved.

**Reply by bluebill1049** [reply]
awesome!

**Comment by Aabhash97** [comment]
I am trying to incorporate multi checkboxes with react select which also shows no of items selected in the dropdown.I can't  use it using react useform Hook as I am unable to register the value using Controller. Below is the code of React Select dropdown (with Select All).
Can Anyone help on this issue
[https://codesandbox.io/s/agitated-golick-dksp6?file=/src/App.js](url)
@bluebill1049 , Any view on this
